src/firefox.py

```diff
@@ -0,0 +1,135 @@
+PRODUCT = "firefox"
+
+DATE_REGEX = r"(January|Jan|February|Feb|March|Mar|April|Apr|May|June|Jun|July|Jul|August|Aug|September|Sept|October|Oct|November|Nov|December|Dec)\s+\d{1,2}(st|nd|rd|th)?,\s+\d{4}"
+VERSION_REGEX = r"\d+(\.\d+)*"
+
+class UnsupportedReleasePageError(Exception):
+    "Raised when a firefox release page is not supported"
+    pass
+
+def format_date(unformatted_date: str) -> str:
+    """ Format date from July 11, 2002 to 2002-07-11 """
+    date = re.sub(r'(\d)(st|nd|rd|th)', r'\1', unformatted_date)
+    formats = ["%b %d, %Y", "%B %d, %Y"]
+    for f in formats:
+        try:
+            return datetime.strptime(date, f).strftime("%Y-%m-%d")
+        except ValueError:
+            pass
+    return ""
+
+def get_version_and_date_gt_28(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 28.0 """
+
+    # get version
+    version = soup.find("div", class_="c-release-version").get_text()
+
+    # get date
+    unformatted_date = soup.find("p", class_="c-release-date").get_text()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date_gt_10(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 10.0 """
+    release_info = soup.find("h2").find("small").text
+
+    # get version
+    version_match = re.search(VERSION_REGEX, soup.select('div#nav-access a')[0].get("href"))
+    if version_match is None:
+        raise UnsupportedReleasePageError("Unable to find version")
+    version = version_match.group()
+
+    # get date
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date_gt_3(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 3.0 """
+    release_info = soup.select('div#main-feature p em')[0].get_text()
+
+    # get version
+    version_match = re.search(VERSION_REGEX, release_info)
+    if version_match is None:
+        raise UnsupportedReleasePageError("Unable to find version")
+    version = version_match.group()
+
+    # get date
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date(soup: BeautifulSoup, release_page: str) -> Tuple[str, str]:
+    """ Get version and date from the given release page """
+    functions = [get_version_and_date_gt_28, get_version_and_date_gt_10, get_version_and_date_gt_3]
+
+    # Note: firefox release pages for versions <3.0 don't include release dates so we
```

Comment:
There is no evidence that it will be added anytime soon, so maybe we could avoid making request for those versions in order to reduce the execution time?

Reply:
Fixed.